include/kernels/madwf_transfer.cuh

```diff
@@ -0,0 +1,150 @@
+#pragma once
+
+#include <color_spinor_field.h>
+#include <color_spinor_field_order.h>
+#include <shared_memory_cache_helper.cuh>
+#include <madwf_transfer.h>
+
+namespace quda
+{
+
+  namespace madwf_ml
+  {
+    constexpr int spin_dim = 4;
+    constexpr int color_dim = 3;
+    constexpr int sm_dim = spin_dim * spin_dim;
+
+    constexpr int color_spin_dim = spin_dim * color_dim;
+    constexpr int wm_dim = color_spin_dim * color_spin_dim;
+
+    template <class real> using WilsonVector = ColorSpinor<real, color_dim, spin_dim>;
+
+    template <class real> using SpinMatrix = Matrix<complex<real>, spin_dim>;
+
+    template <class real, transfer_5D_type transfer_type> struct transfer_5D_mapper {
+    };
+
+    template <class real> struct transfer_5D_mapper<real, transfer_5D_type::Spin> {
+      using type = SpinMatrix<real>;
+    };
+
+    template <bool dagger, class real>
+    __device__ __host__ inline WilsonVector<real> matrix_vector_multiply(const SpinMatrix<real> &m,
+                                                                         const WilsonVector<real> &v)
+    {
+      WilsonVector<real> out; // out is initialized to zero
+#pragma unroll
+      for (int color = 0; color < color_dim; color++) {
+#pragma unroll
+        for (int column = 0; column < spin_dim; column++) {
+          auto v_col = v(column, color);
+#pragma unroll
+          for (int row = 0; row < spin_dim; row++) {
```

Comment:
The code below won't lead to optimal FMA issuance.  Quick fix is to write this as
```c++
if (dagger)
  out(row, color) = cmac(conj(m(column, row), v_col, out(row, color));
else
  out(row, color) = cmac(m(row, column), v_col, out(row, color));
```